diff --git a/arch/mips/include/asm/checksum.h b/arch/mips/include/asm/checksum.h
index 49fcb92..822eadd 100644
--- a/arch/mips/include/asm/checksum.h
+++ b/arch/mips/include/asm/checksum.h
@@ -198,8 +198,9 @@ static inline __sum16 ip_compute_csum(const void *buff, int len)
 }
 
 #define _HAVE_ARCH_IPV6_CSUM
-static __inline__ __sum16 csum_ipv6_unaligned_magic(struct in6_addr *saddr,
-				          struct in6_addr *daddr,
+static __inline__ __sum16 csum_ipv6_unaligned_magic(
+					  const struct in6_addr *saddr,
+				          const struct in6_addr *daddr,
 					  __u32 len, unsigned short proto,
 					  __wsum sum);
 
@@ -271,8 +272,9 @@ static __inline__ __sum16 csum_ipv6_magic(const struct in6_addr *saddr,
 	return csum_fold(sum);
 }
 
-static __inline__ __sum16 csum_ipv6_unaligned_magic(struct in6_addr *saddr,
-				          struct in6_addr *daddr,
+static __inline__ __sum16 csum_ipv6_unaligned_magic(
+					  const struct in6_addr *saddr,
+				          const struct in6_addr *daddr,
 					  __u32 len, unsigned short proto,
 					  __wsum sum)
 {
