diff --git a/include/net/tcp.h b/include/net/tcp.h
index 2d80c29..2d9c42f 100644
--- a/include/net/tcp.h
+++ b/include/net/tcp.h
@@ -655,7 +655,7 @@ struct tcp_skb_cb {
 #define TCPCB_RETRANS		(TCPCB_SACKED_RETRANS|TCPCB_EVER_RETRANS)
 
 	__u32		ack_seq;	/* Sequence number ACK'd	*/
-};
+} __packed;
 
 #define TCP_SKB_CB(__skb)	((struct tcp_skb_cb *)&((__skb)->cb[0]))
 
diff --git a/include/net/udp.h b/include/net/udp.h
index e39592f..757d0b0 100644
--- a/include/net/udp.h
+++ b/include/net/udp.h
@@ -47,7 +47,7 @@ struct udp_skb_cb {
 	} header;
 	__u16		cscov;
 	__u8		partial_cov;
-};
+} __packed;
 #define UDP_SKB_CB(__skb)	((struct udp_skb_cb *)((__skb)->cb))
 
 /**
